capsules/src/sdcard.rs

```diff
@@ -0,0 +1,1424 @@
+    CMD55 = 55, // Next command will be application specific
+    CMD58 = 58, // Read operation condition register (OCR)
+    ACMD41 = 0x80 + 41, // App-specific Init
+}
+
+#[allow(dead_code)]
+#[derive(Clone,Copy,Debug,PartialEq)]
+enum SDResponse {
+    R1,
+    R2,
+    R3,
+    R7,
+}
+
+#[derive(Clone,Copy,Debug,PartialEq)]
+enum State {
+    Idle,
+
+    SendACmd { acmd: SDCmd, arg: u32 },
+
+    InitReset,
+    InitCheckVersion,
+    InitRepeatHCSInit,
+    InitCheckCapacity,
+    InitAppSpecificInit,
+    InitRepeatAppSpecificInit,
+    InitRepeatGenericInit,
+    InitSetBlocksize,
+    InitComplete,
+
+    StartReadBlocks { count: u32 },
+    WaitReadBlock,
+    ReadBlockComplete,
+    WaitReadBlocks { count: u32 },
+    ReceivedBlock { count: u32 },
+    ReadBlocksComplete,
+
+    StartWriteBlocks { count: u32 },
+    WriteBlockResponse,
+    WriteBlockBusy,
+    WaitWriteBlockBusy,
+}
+
+#[derive(Clone,Copy,Debug,PartialEq)]
+enum AlarmState {
+    Idle,
+
+    DetectionChange,
+
+    RepeatHCSInit,
+    RepeatAppSpecificInit,
+    RepeatGenericInit,
+
+    WaitForDataBlock,
+    WaitForDataBlocks { count: u32 },
+
+    WaitForWriteBusy,
+}
+
+#[derive(Clone,Copy,Debug,PartialEq)]
+enum ErrorCode {
+    CardStateChanged = -1,
+    InitializationFailure = -2,
+    ReadFailure = -3,
+    WriteFailure = -4,
+    TimeoutFailure = -5,
+}
+
+// SD Card types
+const CT_MMC: u8 = 0x01;
+const CT_SD1: u8 = 0x02;
+const CT_SD2: u8 = 0x04;
+const CT_BLOCK: u8 = 0x08;
+
+// SD Card capsule, capable of being built on top of by other kernel capsules
+pub struct SDCard<'a, A: hil::time::Alarm + 'a> {
+    // Note: the SpiMasterDevice used by the sdcard cannot be shared by other
+    //  capsules. The current implementation of the SD card requires manually
+    //  handling the chip select line, breaking virtualization on that bus
+    spi: &'a hil::spi::SpiMasterDevice,
+    state: Cell<State>,
+    after_state: Cell<State>,
+
+    alarm: &'a A,
+    alarm_state: Cell<AlarmState>,
+    alarm_count: Cell<u8>,
+
+    is_initialized: Cell<bool>,
+    card_type: Cell<u8>,
+
+    detect_pin: Cell<Option<&'static hil::gpio::Pin>>,
+
+    txbuffer: TakeCell<'static, [u8]>,
+    rxbuffer: TakeCell<'static, [u8]>,
+
+    client: Cell<Option<&'static SDCardClient>>,
+    client_buffer: TakeCell<'static, [u8]>,
+    client_offset: Cell<usize>,
+}
+
+impl<'a, A: hil::time::Alarm + 'a> SDCard<'a, A> {
+    pub fn new(spi: &'a hil::spi::SpiMasterDevice,
+               alarm: &'a A,
+               detect_pin: Option<&'static hil::gpio::Pin>,
+               txbuffer: &'static mut [u8],
+               rxbuffer: &'static mut [u8])
+               -> SDCard<'a, A> {
+
+        // ensure that the write buffer is long enough
+        if txbuffer.len() < 515 || rxbuffer.len() < 515 {
+            panic!("Buffers not long enough to store entire SD card block");
+        }
+
+        // initialize buffers
+        for i in 0..txbuffer.len() {
```

Comment:
I code rust like a C developer. I am aware of this flaw and will make every effort to overcome it  : )

I'll make sure I change all the for loops in my code.